packages/salesforcedx-vscode-core/src/orgBrowser/metadataOutlineProvider.ts

```diff
@@ -57,50 +57,53 @@ export class MetadataOutlineProvider
       return Promise.resolve([org]);
     }
 
-    switch (element.type) {
-      case NodeType.Org:
-        element.children = await this.getTypes();
-        break;
-      case NodeType.MetadataType:
-        element.children = await this.getComponents(element);
-        break;
+    if (!element.children) {
+      switch (element.type) {
+        case NodeType.Org:
+          element.setChildren(await this.getTypes(), NodeType.MetadataType);
+          break;
+        case NodeType.Folder:
+        case NodeType.MetadataType:
+          const type = TypeUtils.FOLDER_TYPES.has(element.fullName)
+            ? NodeType.Folder
+            : NodeType.MetadataCmp;
+          element.setChildren(await this.getComponents(element), type);
+          break;
+      }
     }
-    return Promise.resolve(element.children);
+    return Promise.resolve(element.children!);
   }
 
-  public async getTypes(): Promise<BrowserNode[]> {
+  public async getTypes(): Promise<string[]> {
     const username = this.defaultOrg!;
     const typeUtil = new TypeUtils();
     try {
-      const typesList = await typeUtil.loadTypes(username);
-      const nodeList = typesList.map(
-        type => new BrowserNode(type, NodeType.MetadataType)
-      );
-      return nodeList;
+      return await typeUtil.loadTypes(username);
     } catch (e) {
-      const errorNode = new BrowserNode(parseErrors(e), NodeType.EmptyNode);
-      return [errorNode];
+      throw parseErrors(e);
     }
   }
 
-  public async getComponents(
-    metadataType: BrowserNode
-  ): Promise<BrowserNode[]> {
-    const username = this.defaultOrg!;
-    const cmpUtil = new ComponentUtils();
-    const componentsList = await cmpUtil.loadComponents(
-      username,
-      metadataType.label!
-    );
-    const nodeList = componentsList.map(
-      cmp => new BrowserNode(cmp, NodeType.MetadataCmp)
-    );
-    if (nodeList.length === 0) {
-      nodeList.push(
-        new BrowserNode(nls.localize('empty_components'), NodeType.EmptyNode)
-      );
+  public async getComponents(node: BrowserNode): Promise<string[]> {
+    const cmpUtils = new ComponentUtils();
+    try {
+      if (TypeUtils.FOLDER_TYPES.has(node.fullName)) {
```

Comment:
I don't have a strong opinion on this but this could also be a switch where line 103 is the default behavior.